File: evita/dom/regexp.js
// Copyright (c) 2014 Project Vogue. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

/**
 * @struct { name: string, start: number, end: number }
 */
Editor.RegExp.Match = (function() {
  function Match() {
  }
  return Match;
})();

/**
 * @param {*} target
 * @return {?Array.<string>}
 */
global.Editor.RegExp.prototype.exec = function(target) {
  if (target instanceof Range) {
    var range = /** @type {!Range} */(target);
    return range.match(this);
  }

  var string = target.toString();
  var matches = this.exec_(string);
  if (!matches)
    return null;
  var strings = [];
  return matches.map(function(match) {
    return string.substring(match.start, match.end);
  });
};

File: joana/ast/regexp.cc
// Copyright (c) 2016 Project Vogue. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <tuple>

#include "joana/ast/regexp.h"

namespace joana {
namespace ast {

IMPLEMENT_AST_SYNTAX_0(RegExp, AnyCharRegExp, 0)
IMPLEMENT_AST_SYNTAX_1(RegExp, AssertionRegExp, 0, RegExpAssertionKind, kind)
IMPLEMENT_AST_SYNTAX_0(RegExp, CaptureRegExp, 1)
IMPLEMENT_AST_SYNTAX_0(RegExp, CharSetRegExp, 0)
IMPLEMENT_AST_SYNTAX_0(RegExp, ComplementCharSetRegExp, 0)
IMPLEMENT_AST_SYNTAX_0(RegExp, InvalidRegExp, 0)
IMPLEMENT_AST_SYNTAX_0(RegExp, LiteralRegExp, 0)
IMPLEMENT_AST_SYNTAX_0(RegExp, LookAheadRegExp, 1)
IMPLEMENT_AST_SYNTAX_0(RegExp, LookAheadNotRegExp, 1)
IMPLEMENT_AST_SYNTAX_0(RegExp, RepeatRegExp, 2)

//
// OrRegExpSyntax
//
OrRegExpSyntax::OrRegExpSyntax()
    : SyntaxTemplate(std::tuple<>(),
                     SyntaxCode::OrRegExp,
                     Format::Builder().set_is_variadic(true).Build()) {}

OrRegExpSyntax::~OrRegExpSyntax() = default;

//
// RegExpRepeatSyntax
//
RegExpRepeatSyntax::RegExpRepeatSyntax(RegExpRepeatMethod method,
                                       int min,
                                       int max)
    : SyntaxTemplate(std::make_tuple(method, min, max),
                     SyntaxCode::RegExpRepeat,
                     Format::Builder().set_number_of_parameters(3).Build()) {}

RegExpRepeatSyntax::~RegExpRepeatSyntax() = default;

//
// SequenceRegExpSyntax
//
SequenceRegExpSyntax::SequenceRegExpSyntax()
    : SyntaxTemplate(std::tuple<>(),
                     SyntaxCode::SequenceRegExp,
                     Format::Builder().set_is_variadic(true).Build()) {}

SequenceRegExpSyntax::~SequenceRegExpSyntax() = default;

}  // namespace ast
}  // namespace joana
